relay-server/src/services/buffer/envelope_repository/sqlite.rs

```diff
@@ -0,0 +1,662 @@
+                Self::flush_many(&mut self.envelope_store, envelopes).await;
+            }
+
+            self.buffered_envelopes_size = 0;
+        });
+
+        true
+    }
+
+    /// Checks if there's capacity in the store for more envelopes.
+    pub fn has_store_capacity(&self) -> bool {
+        (self.envelope_store.usage() as usize) < self.max_disk_size
+    }
+
+    /// Retrieves the total count of envelopes in the store.
+    pub async fn store_total_count(&self) -> u64 {
+        self.envelope_store
+            .total_count()
+            .await
+            .unwrap_or_else(|error| {
+                relay_log::error!(
+                    error = &error as &dyn Error,
+                    "failed to get the total count of envelopes for the sqlite envelope store",
+                );
+                // In case we have an error, we default to communicating a total count of 0.
+                0
+            })
+    }
+
+    /// Initializes a set of empty [`EnvelopeStack`]s.
+    fn initialize_empty_stacks(&mut self, project_key_pairs: &HashSet<ProjectKeyPair>) {
+        for &project_key_pair in project_key_pairs {
+            let envelope_stack = self.envelope_stacks.entry(project_key_pair).or_default();
+            // When creating an envelope stack during initialization, we assume data is on disk.
+            envelope_stack.check_disk = true;
+        }
+    }
+
+    /// Returns `true` if there are no [`Envelope`]s for the given [`ProjectKeyPair`], false
+    /// otherwise.
+    fn memory_empty(&self, project_key_pair: ProjectKeyPair) -> bool {
+        self.envelope_stacks
+            .get(&project_key_pair)
+            .map_or(true, |e| e.envelopes.is_empty())
+    }
+
+    /// Determines if the number of buffered envelopes is above the spool threshold.
+    fn above_spool_threshold(&self) -> bool {
+        self.buffered_envelopes_size >= self.disk_batch_size as u64
+    }
+
+    /// Spools envelopes to disk, evenly distributing across all project key pairs.
+    ///
+    /// This method attempts to spool `self.disk_batch_size` envelopes to disk, taking an equal
+    /// number from each group of envelopes (represented by different project key pairs) in the
+    /// `self.envelopes` HashMap.
+    ///
+    /// This approach ensures a balanced spooling across all project key pairs, preventing any
+    /// single group from monopolizing the disk storage.
+    async fn spool_to_disk(&mut self) -> Result<(), SqliteEnvelopeRepositoryError> {
+        let mut envelopes_to_spool = Vec::with_capacity(self.disk_batch_size);
+
+        for envelope_stack in self.envelope_stacks.values_mut() {
+            envelopes_to_spool.append(&mut envelope_stack.envelopes);
+            envelope_stack.check_disk = true;
+            relay_statsd::metric!(
+                histogram(RelayHistograms::BufferInMemoryEnvelopesPerKeyPair) =
+                    envelope_stack.envelopes.len() as u64
+            );
+        }
```

Comment:
This looks very similar to the `flush` code, we should unite them or make one call the other.